src/runner/test.rs

```diff
@@ -73,31 +79,76 @@ fn run_cargo<DB: WriteResults>(
     };
 
     let mut did_ice = false;
-    let mut tracker = |line: &str| {
-        did_ice |= line.contains("error: internal compiler error");
+    let mut error_codes = BTreeSet::new();
+    let mut deps = BTreeSet::new();
+
+    let mut detect_error = |line: &str, actions: &mut ProcessLinesActions| {
+        // Avoid trying to deserialize non JSON output
+        if line.starts_with('{') {
+            let message = serde_json::from_str(line);
+            if let Ok(message) = message {
+                match message {
+                    Message::CompilerMessage(compiler_message) => {
+                        let inner_message = compiler_message.message;
+                        match (
+                            inner_message.level,
+                            Crate::try_from(&compiler_message.package_id),
+                        ) {
+                            // the only local crate in a well defined job is the crate currently being tested
```

Comment:
I'd prefer if we fetch all the package IDs belonging to the current crate from cargo metadata, and see if the package ID matches the ones we know of. I'm pretty sure we're fetching the cargo metadata somewhere already.

This way the check is more accurate, and we can also avoid parsing the package ID a lot of the times (since we can just compare strings for messages originating in the same crate).